api/pkg/oauth/auth/service/auth.go

```diff
@@ -0,0 +1,193 @@
+	"gorm.io/gorm"
+)
+
+type service struct {
+	app.Service
+	api app.Config
+}
+
+type request struct {
+	db            *gorm.DB
+	log           *log.Logger
+	defaultScopes []string
+	jwtConfig     *app.JWTConfig
+}
+
+type AuthService struct {
+	Name   string `json:"name"`
+	Status string `json:"status"`
+}
+
+type Services struct {
+	Service AuthService `json:"services"`
+}
+
+var (
+	UI_URL string
+)
+
+type Service interface {
+	AuthCallBack(res http.ResponseWriter, req *http.Request)
+	HubAuthenticate(res http.ResponseWriter, req *http.Request)
+}
+
+// New returns the auth service implementation.
+func New(api app.Config) Service {
+	return &service{
+		Service: api.Service("auth"),
+		api:     api,
+	}
+}
+
+// Return name and status of the services
+func Status(res http.ResponseWriter, req *http.Request) {
+
+	authSvc := Services{
+		AuthService{
+			Name:   "auth",
+			Status: "ok",
+		},
+	}
+
+	var log log.Logger
+	res.WriteHeader(http.StatusOK)
+	if err := json.NewEncoder(res).Encode(authSvc); err != nil {
+		log.Error(err)
+	}
+}
+
+// Authenticates user with the speicified git provider
+// using goth and calls the AuthCallback function
+func Authenticate(res http.ResponseWriter, req *http.Request) {
+	UI_URL = req.FormValue("redirect_uri")
+	gothic.BeginAuthHandler(res, req)
+}
+
+// Once user is authenticated, store the user details in db
+// and redirect to UI with the status code and auth code
+func (s *service) AuthCallBack(res http.ResponseWriter, req *http.Request) {
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	ghUser, err := gothic.CompleteUserAuth(res, req)
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	params := req.URL.Query()
+
+	if err = r.insertData(ghUser, params.Get("code")); err != nil {
+		r.log.Error(err)
+		res.Header().Set("Location", fmt.Sprintf("%s?status=%d", UI_URL, http.StatusBadRequest))
+		res.WriteHeader(http.StatusTemporaryRedirect)
+	}
+
+	res.Header().Set("Location", fmt.Sprintf("%s?status=%d&code=%s", UI_URL, http.StatusOK, params.Get("code")))
+	res.WriteHeader(http.StatusTemporaryRedirect)
+}
+
+// Checks for auth code in the headers, validates the
+// auth code and returns the jwt token for user
+func (s *service) HubAuthenticate(res http.ResponseWriter, req *http.Request) {
+
+	// Get the auth code from params
+	code := req.FormValue("code")
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	var User model.User
+	// Check if user exist
+	q := r.db.Model(&model.User{}).
+		Where("code = ?", code)
+
+	err := q.First(&User).Error
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusBadRequest)
```

Comment:
if user not found then it would be  bad req I guess
we should handle not found error explicitly

Reply:
how we can define the status error explicitly ??

Reply:
if not found err { 
  bad req
}
internal err